### PR TITLE
Customize load balanced requests according to the chosen ServiceInstance

### PR DESCRIPTION
**Pull request by galaxy-sea**
Applications can define their own LoadBalancerFeignRequestTransformer beans
which can modify the Request to be executed. These beans can be
@Ordered in case of multiple transformers.

reference [Transform the load-balanced HTTP request]( https://docs.spring.io/spring-cloud-commons/docs/current/reference/html/#transform-the-load-balanced-http-request)

### CODE REVIEW

**Comment by galaxy-sea** [comment]
@OlgaMaciaszek 
Can you check this PR.
I hope this function can be seen in the next release.

**Comment by OlgaMaciaszek** [comment]
Hello, @galaxy-sea Thanks for submitting. Will review this week.

File: spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/loadbalancer/LoadBalancerFeignRequestTransformer.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ */
+@Order(LoadBalancerFeignRequestTransformer.DEFAULT_ORDER)
+public interface LoadBalancerFeignRequestTransformer {
+
+	/**
+	 * Order for the {@link LoadBalancerFeignRequestTransformer}.
+	 */
+	int DEFAULT_ORDER = 0;
+
+	Request transformRequest(Request request, ServiceInstance instance);
```
Comment (OlgaMaciaszek) [inline spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/loadbalancer/LoadBalancerFeignRequestTransformer.java:38]: Please add javadoc for method.

**Review by OlgaMaciaszek — CHANGES_REQUESTED** [review]
Thanks for submitting the PR @galaxy-sea. In general, LGTM. There's just one javadoc missing - have added a comment, please address. Also, have requested @Buzzardo to review the documentation change.

**Comment by galaxy-sea** [comment]
Hello @OlgaMaciaszek, I have submitted the Javadoc

File: spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/loadbalancer/LoadBalancerFeignRequestTransformer.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+@Order(LoadBalancerFeignRequestTransformer.DEFAULT_ORDER)
+public interface LoadBalancerFeignRequestTransformer {
+
+	/**
+	 * Order for the {@link LoadBalancerFeignRequestTransformer}.
+	 */
+	int DEFAULT_ORDER = 0;
+
+	/**
+	 * ServiceInstance to transform the load-balanced Request.
```
Comment (OlgaMaciaszek) [inline spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/loadbalancer/LoadBalancerFeignRequestTransformer.java:39]: Should be sth more about the method like "Allows transforming load-balanced requests based on the provided {@link ServiceInstance}`

**Review by OlgaMaciaszek — CHANGES_REQUESTED** [review]
Please edit the javadoc.

File: spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/loadbalancer/LoadBalancerFeignRequestTransformer.java (hunk 2 above)
Comment (galaxy-sea) [reply spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/loadbalancer/LoadBalancerFeignRequestTransformer.java]: Thank, I have re edited Javadoc

**Review by OlgaMaciaszek — APPROVED** [review]
Thanks @galaxy-sea. This look good to me now. Once @Buzzardo reviews the doc changes and any necessary fixes are made there, we'll be able to merge it.

File: docs/src/main/asciidoc/spring-cloud-openfeign.adoc [hunk 3]
```diff
@@ -815,6 +815,33 @@ Sometimes, when load balancing is enabled for Feign clients, you may want to use
...
+				headers.put("X-ServiceId", Collections.singletonList(instance.getServiceId()));
+				headers.put("X-InstanceId", Collections.singletonList(instance.getInstanceId()));
+				return Request.create(request.httpMethod(), request.url(), headers, request.body(), request.charset(),
+						request.requestTemplate());
+			}
+		};
+	}
+----
+
+If multiple transformers are defined, they are applied in the order in which Beans are defined.
```
Comment (Buzzardo) [inline docs/src/main/asciidoc/spring-cloud-openfeign.adoc:842]: Uncapitalize "Beans".

File: docs/src/main/asciidoc/spring-cloud-openfeign.adoc [hunk 4]
```diff
@@ -815,6 +815,33 @@ Sometimes, when load balancing is enabled for Feign clients, you may want to use
 spring.cloud.openfeign.oauth2.load-balanced=true
 ----
 
+=== Transform the load-balanced HTTP request
+
+You can use the selected `ServiceInstance` to transform the load-balanced HTTP Request.
+
+For `Request`, you need to implement and define `LoadBalancerFeignRequestTransformer` as follows:
```
Comment (Buzzardo) [inline docs/src/main/asciidoc/spring-cloud-openfeign.adoc:822]: Add a comma before "as follows".

**Review by Buzzardo — CHANGES_REQUESTED** [review]

**Comment by galaxy-sea** [comment]
hello @Buzzardo , I have re edited spring-cloud-openfeign.adoc.

**Review by Buzzardo — APPROVED** [review]

**Comment by Buzzardo** [comment]
Thank you. I approved the changes.

**Comment by galaxy-sea** [comment]
Thank @OlgaMaciaszek @Buzzardo

**Comment by OlgaMaciaszek** [comment]
Thanks for the contribution @galaxy-sea

**Pull request merged**